
# HG changeset patch
# User Geoff Lankow <geoff@darktrojan.net>
# Date 1644007033 -3600
# Node ID d34feb5081ce7c23a514a74a376d5efe3df1fbf2
# Parent  6484b56f1acc0c1e1c2e56adf6beacf4c2e85f8c
Bug 1747360 - Catch error opening attachment with an unknown content type. r=aleca
For some reason deep in the parser code, application/octet-stream attachments lose their content type.
This causes the new attachment opening code to throw an exception on Windows if nothing is registered for the attachment's extension.
If an exception is thrown here (for this reason or any other) fall back to the generic application/octet-stream type.

Differential Revision: https://phabricator.services.mozilla.com/D135664

diff --git a/mail/base/content/msgHdrView.js b/mail/base/content/msgHdrView.js
--- a/mail/base/content/msgHdrView.js
+++ b/mail/base/content/msgHdrView.js
@@ -1956,20 +1956,30 @@ AttachmentInfo.prototype = {
           this.uri,
           this.isExternalAttachment
         );
         return;
       }
 
       // Get the MIME info from the service.
 
-      let mimeInfo = gMIMEService.getFromTypeAndExtension(
-        this.contentType,
-        extension
-      );
+      let mimeInfo;
+      try {
+        mimeInfo = gMIMEService.getFromTypeAndExtension(
+          this.contentType,
+          extension
+        );
+      } catch (ex) {
+        // If the call above fails, which can happen on Windows where there's
+        // nothing registered for the file type, assume this generic type.
+        mimeInfo = gMIMEService.getFromTypeAndExtension(
+          "application/octet-stream",
+          null
+        );
+      }
       // The default action is saveToDisk, which is not what we want.
       // If we don't have a stored handler, ask before handling.
       if (!gHandlerService.exists(mimeInfo)) {
         mimeInfo.alwaysAskBeforeHandling = true;
         mimeInfo.preferredAction = Ci.nsIHandlerInfo.alwaysAsk;
       }
 
       // If we know what to do, do it.
